### PR TITLE
feat: execute blocks starting from genesis

### LINKED ISSUE #218: Add extrinsics validation
- Implement transaction queue
- Implement [validate transactions and store algorithm](https://spec.polkadot.network/chap-state#algo-validate-transactions)
- Implement [maintain transaction pool algorithm](https://spec.polkadot.network/chap-state#algo-maintain-transaction-pool)

### LINKED ISSUE #225: Add test to showcase successful block building(full node)

### LINKED ISSUE #293: Full sync state machine
State machine state:
- RequestBlockInfoState
- ValidatePreState
  - Check_inherents
- ExecuteBlock
- ValidatePostState
  - Validate merkle root

### LINKED ISSUE #296: BlockExecutor to Runtime to HostAPI logic
Image

### PR DESCRIPTION
**Pull request by ablax**
# Description

- What does this PR do?
Executes blocks starting from genesis (tested on kusama and polkadot)
- Why are these changes needed?
It is an essential part of full blockchan node
- How were these changes implemented and what do they affect?
Currently the changes break old tests, but they will be updated accordingly
Also some debug coments with log.fine are left, but I think it is not a bad idea to have such for debugging in future.
They are disabled by default.

Fixes #293

## Checklist:
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [x] My PR title matches the [Conventional Commits spec](https://www.conventionalcommits.org/).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/main/java/com/limechain/storage/trie/TrieStorage.java [hunk 1]
```diff
@@ -408,76 +382,83 @@ private boolean startsWith(byte[] array, byte[] prefix) {
      * @return A list of pairs, each containing a set of nibbles (representing a key within the trie)
      * and the corresponding node data.
      */
-    public List<StorageNode> entriesBetween(byte[] merkleValue, String searchKey) {
+    // FIXME: Epic mismatch: this function returns all the nodes on the paths; but the caller expects storage nodes only.
+    //  Also, when talking about a partial trie, we can't simply insert the storage nodes only (i.e. the storage KVPs)
+    //  since the partial trie won't have any information about the entire trie it's a part of, so that'll offset merkle values and stuff
+    //  Full structural replication will be necessary, but the TrieStructure isn't fit for that purpose :D
```
Comment (georg-getz) [inline src/main/java/com/limechain/storage/trie/TrieStorage.java]: Is this comment outdated?

File: src/main/java/com/limechain/storage/trie/TrieStorage.java [hunk 2]
```diff
@@ -408,76 +382,83 @@ private boolean startsWith(byte[] array, byte[] prefix) {
...
+                return;
+            }
 
-        if (childrenMerkleValues.size() < key[commonPrefix]) {
-            return;
+            childIndexWithinParent = keyIter.next();
         }
-        byte[] childMerkleValue = childrenMerkleValues.get(key[commonPrefix]);
+
+
```
Comment (georg-getz) [inline src/main/java/com/limechain/storage/trie/TrieStorage.java]: CEL

File: src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java [hunk 3]
```diff
@@ -13,6 +35,94 @@ public FullSyncMachine(final Network networkService) {
...
+        this.networkService.updateCurrentSelectedPeerWithNextBootnode();
+        AccessorHolder.getInstance().setToGenesis(); //todo: dirty fix for now
+
+        int startNumber = 1;
+        List<SyncMessage.BlockData> receivedBlockDatas = getBlocks(startNumber, 10);
+        while (!receivedBlockDatas.isEmpty()) {
+            startNumber += 10;
+            executeBlocks(receivedBlockDatas);
+            receivedBlockDatas = getBlocks(startNumber, 10);
+        }
```
Comment (georg-getz) [inline src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java]: How was the number 10 chosen?

File: src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java [hunk 4]
```diff
@@ -13,6 +35,94 @@ public FullSyncMachine(final Network networkService) {
...
+            ));
+
+            // Get the block from the list of responses:
+            return response.getBlocksList();
+        } catch (Exception ex) {
+            log.info("Error while fetching blocks, trying to fetch again");
+            this.networkService.updateCurrentSelectedPeerWithNextBootnode();
+            return getBlocks(start, amount);
+        }
+    }
```
Comment (georg-getz) [inline src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java]: I think we usually put the annotation on top of the function.

File: src/main/java/com/limechain/sync/fullsync/VerifyBlockState.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.limechain.sync.fullsync;
+
+public class VerifyBlockState implements FullSyncState {
+    @Override
+    public FullSyncState transition(FullSyncMachine fullSyncMachine) {
+        return null;
+    }
+}
```
Comment (georg-getz) [inline src/main/java/com/limechain/sync/fullsync/VerifyBlockState.java]: Are we using this states anywhere? I thought David had reached the conclusion that the full sync won't require states unlike the warp sync.

File: src/main/java/com/limechain/sync/fullsync/inherents/InherentData.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Represents the totality of inherentes included in each block.
+ * @param timestamp Number of milliseconds since the UNIX epoch when the block is generated, ignoring leap seconds.
+ *                  Its identifier passed to the runtime is: `timstap0`.
+ * @see <a href="https://spec.polkadot.network/chap-state#tabl-inherent-data">the spec</a>
+ */
+public record InherentData(
+    long timestamp
+    // TODO: parachain-related inherents are missing
```
Comment (georg-getz) [inline src/main/java/com/limechain/sync/fullsync/inherents/InherentData.java]: If those inherents are important only for relaying nodes you might delete this comment as that is not happening anytime soon.

File: src/main/java/com/limechain/trie/TrieAccessor.java [hunk 7]
```diff
@@ -135,35 +81,35 @@ public List<byte[]> findKeysByPrefix(Nibbles prefixSeek, int limit) {
...
-                .map(parentKey -> prefixIndexInParent(parentKey, prefix))
-                .flatMap(parent::getChild)
-                .map(c -> lexicographicDelete(c, limit))
-                .orElse(new DeleteByPrefixResult(0, true));
+//        NodeHandle<NodeData> parent = switch (partialTrie.node(prefix)) {
+//            case Vacant<NodeData> vacant -> Optional.ofNullable(vacant.getClosestAncestorIndex())
+//                    .map(partialTrie::nodeHandleAtIndex)
+//                    .orElse(null);
+//            case NodeHandle<NodeData> handle -> handle.getParent();
+//        };
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/TrieAccessor.java]: forgotten?

File: src/main/java/com/limechain/trie/TrieAccessor.java [hunk 8]
```diff
@@ -135,35 +81,35 @@ public List<byte[]> findKeysByPrefix(Nibbles prefixSeek, int limit) {
...
 
+        DeleteByPrefixResult other = new DeleteByPrefixResult(0, true);
+        return other;
+//        return Optional.ofNullable(parent)
+//                .map(NodeHandle::getFullKey)
+//                .map(parentKey -> prefixIndexInParent(parentKey, prefix))
+//                .flatMap(parent::getChild)
+//                .map(c -> lexicographicDelete(c, limit))
+//                .orElse(other);
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/TrieAccessor.java]: forgotten?

File: src/main/java/com/limechain/trie/TrieDiff.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import com.limechain.trie.structure.nibble.Nibbles;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+public class TrieDiff {
+    private final NavigableMap<Nibbles, Boolean> availability = new TreeMap<>(Comparator.comparing(Nibbles::toString));
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/TrieDiff.java]: you can add a short description of what the class exists for above the class definition

File: src/main/java/com/limechain/trie/TrieDiff.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        other.difference.forEach((key, value) -> {
+            this.difference.put(key, value);
+            this.availability.put(key, true);
+        });
+        other.availability.forEach(this.availability::putIfAbsent);
+    }
+
+    public Nibbles storageNextKey(Nibbles key, Nibbles inParentNextKey, boolean orEqual) {
+        return null; // Placeholder for the actual logic.
+    }
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/TrieDiff.java]: Is this intentionally left so?

File: src/main/java/com/limechain/trie/structure/StorageNodeHandle.java [hunk 11]
```diff
@@ -13,4 +13,16 @@ public final class StorageNodeHandle<T> extends NodeHandle<T> {
     public boolean hasStorageValue() {
         return true;
     }
+
+    public boolean clearStorageValue() {
+        return trieStructure.clearNodeValue(rawNodeIndex);
+    }
+
+    //convert to branchNode
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/structure/StorageNodeHandle.java]: you can delete this, no point in commenting if the function name is going to give the same info

File: src/main/java/com/limechain/trie/structure/TrieStructure.java [hunk 12]
```diff
@@ -524,9 +519,20 @@ private boolean clearNodeValue(int nodeIndex) {
...
+                .filter(Objects::nonNull)
+                .count();
+
         if (numberOfChildren == 0) {
             parentNode.childrenIndices[parent.childIndexWithinParent().asInt()] = null;
             nodes.remove(nodeIndex);
+
+            if(parentChildren == 1 && !parentNode.hasStorageValue) {
+                return clearNodeValue(parent.parentNodeIndex());
+            }else if(parentChildren == 2){
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/structure/TrieStructure.java]: spacing

File: src/main/java/com/limechain/network/protocol/blockannounce/scale/BlockScaleWriter.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+@NoArgsConstructor(access  = AccessLevel.PRIVATE)
+public class BlockScaleWriter implements ScaleWriter<Block> {
+    private static final BlockScaleWriter INSTANCE = new BlockScaleWriter();
+
+    @Override
+    public void write(ScaleCodecWriter scaleCodecWriter, Block block) throws IOException {
+
+    }
+}
```
Comment (georg-getz) [inline src/main/java/com/limechain/network/protocol/blockannounce/scale/BlockScaleWriter.java]: delete this file

**Review by georg-getz — APPROVED** [review]

File: src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java (hunk 3 above)
Comment (ablax) [reply src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java]: It was chosen while testing, currently set to 50, not sure what would be the most appropriate number

**Pull request merged**